File: Cplusplus/0090_subsets_ii/0090_subsets_ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> list;
        if (nums.size() == 0) {
            res.push_back(list);
            return res;
        }
        sort(nums.begin(), nums.end());
        backtracking(res, list, nums, 0);
        return res; 
    }
    
    void backtracking(vector<vector<int>> &res, vector<int> list, vector<int> &nums, int pos) {
        res.push_back(list);    
        for (int i = pos; i < nums.size(); i++) {
            /* ignore traversed situation though i != pos and nums duplication. */
            if (i > 0 && nums[i] == nums[i - 1] && i != pos) 
                continue;
            list.push_back(nums[i]);
            backtracking(res, list, nums, i + 1);
            list.pop_back();
        }
    }
};

File: Java/0090_subsets_ii/0090_subsets_ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        ArrayList solution = new ArrayList<ArrayList<Integer>>();
        if (nums.length == 0) {
            solution.add(list);
            return solution;
        }     
        Arrays.sort(nums);     
        backtracking(solution, list, nums, 0);        
        return solution;
    }
    
    public static void backtracking(ArrayList<ArrayList<Integer>> solution, 
                                    ArrayList<Integer> list, int[] nums, int pos) {
        solution.add(new ArrayList<Integer>(list));     
        for (int i = pos; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1] && i != pos)
                continue;
            list.add(nums[i]);
            backtracking(solution, list, nums, i + 1);
            list.remove(list.size() - 1);
        }
    }
}